Update to hickory-dns 0.25.1

remilia-dev opened the pull request:
# Changes
* Updates hickory crates to `0.25.1`
* Updates `rustls` to `0.23`
* Updates `rustls-pemfile` to `2.2.0`
* Adds the ability to build with `ring` or `aws-lc-rs`
  * ~~These are currently mutually exclusive (either you build with the `ring` provider or the `aws-lc-rs` provider, but not both)~~
  * If both `aws-lc-rs` and `ring` are enabled, `ring` is used to match [hickory's behavior](https://github.com/hickory-dns/hickory-dns/blob/0f44977800ca8954e86723fcc391514008d2bb1d/crates/proto/src/rustls/mod.rs#L50).

# Not Included
* This does not make the blocklist utilize hickory's `blocklist` feature

# Tested & Working
* udp, tls, & https upstreams
  * This was done with both `native-roots` and `webpki-roots` configured (separate test runs)
* udp, tls, https, & quic downstreams
  * This was done using a real cert with both `ring` and `aws-lc-rs` configured (separate test runs)
* `cargo test`
* `act -j test` (running the test CI locally)

msrd0 commented on Cargo.toml:
```diff
@@ -16,9 +16,11 @@ include = [
 [features]
-default = ["native-roots"]
-native-roots = ["hickory-resolver/native-certs", "reqwest/rustls-tls-native-roots"]
+default = ["native-roots", "ring"]
```
Is there a particular reason we want ring to be the default? Iirc hickory uses aws-lc as default, and aws-lc has the advantage of supporting more architectures than ring.

@LuckyTurtleDev do you have an opinion on this?

remilia-dev replied:
I had gone with ring since `dnssec-ring` was previously used but your points make sense (especially if hickory is using it by default). Latest commit switches the default to `aws-lc-rs`.

msrd0 commented on src/main.rs:
```diff
@@ -10,6 +10,9 @@ extern crate test;
 static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
 
+#[cfg(all(feature = "ring", feature = "aws-lc-rs"))]
+compile_error! {"Feature \"ring\" and  feature \"aws-lc-rs\" cannot be enabled at the same time."}
```
Is this something we have to do or does rustls/hickory already do it for us?

Also, do we even need this? It looks like the way the code is written right now, aws-lc would overwrite but not conflict ring.

msrd0 replied:
In this case (since crab-hole is a binary) you are right, there is not really a point in having both enabled.

However, when you write a library, let's call it _a_, and some other library, let's call it _b_, uses library _a_, it can be useful to have features that "additionally" "disable" the default for _a_ so that eventhough library _b_ forces the default features of library _a_, you can have the alternative behaviour.

msrd0 commented on src/main.rs:
```diff
@@ -167,7 +172,23 @@ impl RequestHandler for Handler {
+			debug!(
+				"Failed to get request info from request with id {}",
+				request.id()
+			);
```
I'm not sure I agree with this being a `debug` message

msrd0 commented on src/main.rs:
```diff
@@ -213,37 +234,36 @@ fn reader_for(path: &Path) -> anyhow::Result<BufReader<File>> {
+			{
+				Ok(s) => Some(s),
+				Err(_) => None
+			}
```
Right here, we construct a `Result` with context just to throw it away and make it an option (which btw can be done by calling `.ok()`). I think the best option here would be to issue a `warn` log message with the error.

msrd0 commented on src/main.rs:
```diff
@@ -213,37 +234,36 @@ fn reader_for(path: &Path) -> anyhow::Result<BufReader<File>> {
+		key,
+		CryptoProvider::get_default().ok_or_else(|| {
+			anyhow!("CryptoProvider default should have been registered in main!")
+		})?
```
I think this could just as well be a `.expect()`. There is no way to recover a missing crypto provider.

msrd0 reviewed (changes requested):
Thanks for the PR. I have a few comments about details.

msrd0 commented:
Also, please upload your GPG key to github so your commits don't show as _Unverified_.

remilia-dev commented:
> Also, please upload your GPG key to github so your commits don't show as _Unverified_.

Woops! Looks like I hadn't added the subkey to GitHub yet.

I've pushed up a commit that addresses your feedback. Let me know if any further adjustments need to be made!

msrd0 reviewed (approved):
LGTM, I'm going to wait before merging to see if @LuckyTurtleDev wants to add a review.

msrd0 commented:
Hm now we need to teach CI that it always has to enable one of the features ...

remilia-dev commented:
> Hm now we need to teach CI that it always has to enable one of the features ...

Unfortunately it doesn't look like there's a nice solution with `cargo-all-features`. It does not currently have a way to always include a feature from a group of features (there's an [open pull request](https://github.com/frewsxcv/cargo-all-features/pull/44) for that).

`cargo-hack` does support this, so I've pushed a commit adjusting the CI file to utilizing `cargo hack test --feature-powerset --at-least-one-of aws-lc-rs,ring --locked -- --include-ignored`

I did a test-run of the `test` job locally using [act](https://github.com/nektos/act) after going through 32 feature permutations.

## Combined `ring` and `aws-lc-rs` Behavior
On inspection it looks like specifying `hickory-server/tls-ring` causes `ring` to be the default over `aws-lc-rs` ([see here](https://github.com/hickory-dns/hickory-dns/blob/0f44977800ca8954e86723fcc391514008d2bb1d/crates/proto/src/rustls/mod.rs#L50)). To avoid hickory using one crypto provider while certificate loading uses another, I've changed it to where having both `ring` and `aws-lc-rs` enabled will cause `ring` to be utilized.

msrd0 commented:
Thanks, that sounds good. But I do wonder why 32 permutations? The most that include either aws-lc or ring should be 24, and when we exclude the `__debug_parser` feature it would only be 12, which sounds a lot more reasonable.

LuckyTurtleDev commented:
Thanks for pr  and sorry for the long waiting time.
